File: server/Services/KeepsService.cs




namespace Keepr.Services;

public class KeepsService
{
    private readonly KeepsRepository _repository;

    public KeepsService(KeepsRepository repository)
    {
        _repository = repository;
    }

    internal Keep CreateKeep(Keep keepData)
    {
        Keep keep = _repository.CreateKeep(keepData);
        return keep;
    }

    internal string DeleteKeep(int keepId, string userId)
    {
        Keep keep = GetKeepById(keepId);
        if (keep.CreatorId != userId)
        {
            throw new Exception("Not your keep to delete!");
        }
        _repository.DeleteKeep(keepId);
        return $"{keep.Name} has been deleted!";
    }

    internal Keep EditKeep(int keepId, string userId, Keep keepData)
    {
        Keep ogKeep = GetKeepById(keepId);
        if (ogKeep.CreatorId != userId)
        {
            throw new Exception($"Not your keep to edit {ogKeep.Name}");
        }

        ogKeep.Name = keepData.Name ?? ogKeep.Name;
        ogKeep.Img = keepData.Img ?? ogKeep.Img;
        ogKeep.Description = keepData.Description ?? ogKeep.Description;

        Keep keep = _repository.EditKeep(ogKeep);
        return keep;
    }

    internal Keep GetKeepById(int keepId)
    {
        Keep keep = _repository.GetKeepById(keepId);
        if (keep == null)
        {
            throw new Exception($"invalid Id:{keepId}");
        }
        return keep;
    }

    internal List<Keep> GetKeeps()
    {
        List<Keep> keeps = _repository.GetKeeps();
        return keeps;
    }
}

File: client/src/services/KeepsService.js
import { api } from "./AxiosService";
import { logger } from '../utils/Logger'
import { AppState } from "../AppState";
import { Keep } from "../models/Keep";

class KeepsService {
    async getKeeps() {
        const res = await api.get("api/keeps")
        AppState.keeps = res.data.map(keepPojo => new Keep(keepPojo))
        logger.log("this is the keeps in AS", AppState.keeps)

    }

    async getKeepById(keepId) {
        const res = await api.get(`api/keeps/${keepId}`)
        AppState.activeKeep = res.data
        logger.log(res.data)
    }

    async createKeep(keepData) {
        const res = await api.post(`api/keeps`, keepData)
        const newKeep = new Keep(res.data)
        AppState.keeps.push(newKeep)
        return newKeep
    }

    async deleteKeep(keepId) {
        const res = await api.delete(`api/keeps/${keepId}`)
        logger.log("keeps has been deleted", res.data)
        AppState.keeps = AppState.keeps.filter((keep) => keep.id != keepId)
    }
}

export const keepsService = new KeepsService()